lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +279,140 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
     build (AddressPool acct (AddressPoolGap g) _) = mempty
         <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
       where
-        ccF = build $ toText $ accountingStyle @chain
+        ccF = build $ toText $ role @chain
+        xpubF = hexF $ serializeXPub acct
+        acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
+
+data VerificationKeyPool (key :: Depth -> * -> *) = VerificationKeyPool {
+      verPoolAccountPubKey
+        :: !(key 'AccountK XPub)
+    -- ^ Corresponding key for the pool (a pool is tied to only one account)
+    , verPoolGap
+        :: !AddressPoolGap
+    -- ^ The actual gap for the pool. This can't change for a given pool.
+    , verPoolIndexedKeys
+        :: !(Map KeyHash (Index 'Soft 'ScriptK, AddressState))
+    , verPoolKnownScripts :: !(Map ScriptHash [Index 'Soft 'ScriptK])
+        -- ^ Known script hashes that contain our verification key hashes
+        -- represented here by corresponding indices
+    } deriving (Generic)
+
+deriving instance (Show (key 'AccountK XPub), Show (key 'ScriptK XPub))
+    => Show (VerificationKeyPool key)
+
+deriving instance (Eq (key 'AccountK XPub), Eq (key 'ScriptK XPub))
+    => Eq (VerificationKeyPool key)
+
+instance (NFData (key 'AccountK XPub), NFData (key 'ScriptK XPub))
+    => NFData (VerificationKeyPool key)
+
+instance ((PersistPublicKey (key 'AccountK)))
+    => Buildable (VerificationKeyPool key) where
+    build (VerificationKeyPool acct (AddressPoolGap g) _ _) = mempty
+       <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
+      where
+        ccF = build $ toText MultisigScript
         xpubF = hexF $ serializeXPub acct
         acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
 
--- | Bring a 'AccountingStyle' type back to the term-level. This requires a type
+deriving instance Ord KeyHash
+
+deriveKeyHash
+    :: (WalletKey k, SoftDerivation k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+deriveKeyHash accXPub' =
+    toVerKeyHash . deriveAddressPublicKey accXPub' MultisigScript . coerce
+
+lookupKeyHash
+    :: KeyHash
+    -> VerificationKeyPool k
+    -> (Maybe (Index 'Soft 'ScriptK), VerificationKeyPool k)
+lookupKeyHash keyHash pool@(VerificationKeyPool accXPub g indexedHashKeys scripts) =
+    let updateKey (ix, _) = (ix, Used)
+    in case Map.lookup keyHash indexedHashKeys of
+        Nothing -> (Nothing, pool)
+        Just (ix, _) ->
+            ( Just ix
+            , VerificationKeyPool accXPub g
+                (Map.adjust updateKey keyHash indexedHashKeys) scripts)
```

Comment:
In principle, this function should also make sure to extend the pool if we are looking up a key hash near the "edge" (that is, one of the last `gap` consecutive previously unused addresses)

Reply:
indeed -> https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/2ffef82dd2538db1123289ca95079056fca48894

Reply:
Note that `lookup + adjust` can be replaced with: `alterF` using `(a,)` as a Functor. 

--- 

So, if I understand the strategy here, it is to reconstruct a new pool after having looked up and found a key assuming that `unsafeVerificationKeyPool` does enforce the invariant then it should work indeed but I fear may not be efficient (see next comment) without a required change. Few questions / remarks however:

a) If `unsafeVerificationKeyPool` is effectively maintaining the invariant... Does it really deserve its `unsafe` prefix :) ? When I originally suggested this function name, I had in mind a very simple wrapper around the constructor to be used in places like the database unserialization function (to avoid exposing the constructor but still allow constructing from raw piece of data). It seems like you've taken another direction in the end, but now it makes me wonder: is there really anything unsafe about this constructor anymore? 

b) I assume you only really want to re-construct a new pool if and only if the key was `unused`. Turning a Used into Used should not change anything and we could save a bit of CPU by returning the same pool.

Reply:
ad a) you are write I have make unsafeVerificationKeyPool really unsafe ;-)
ad b) https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/32ce104f76d3cf1efd573bc8dee37b3508334df5